File: javascript/src/agents/judge/judge-span-collector.ts
import { SpanProcessor, ReadableSpan, Span } from "@opentelemetry/sdk-trace-base";
import { attributes } from "langwatch/observability";

/**
 * Collects OpenTelemetry spans for judge evaluation.
 * Implements SpanProcessor to intercept spans as they complete.
 */
export class JudgeSpanCollector implements SpanProcessor {
  private spans: ReadableSpan[] = [];
  /**
   * Every span id this process ever STARTED, keyed by trace id. The remote
   * trace fetcher uses it to recognize the scenario process's own spans when
   * the platform echoes them back. The per-thread view below cannot serve
   * that: it walks ancestor attributes, and the walk breaks on spans whose
   * ancestor chain crosses a still-open span (the current turn's spans) or
   * whose instrumentation never tags the thread id (the judge's own model
   * calls via instrumented SDKs).
   */
  private processSpanIds = new Map<string, Set<string>>();

  onStart(span: Span): void {
    const ctx = span.spanContext();
    let ids = this.processSpanIds.get(ctx.traceId);
    if (!ids) {
      ids = new Set();
      this.processSpanIds.set(ctx.traceId, ids);
    }
    ids.add(ctx.spanId);
  }

  onEnd(span: ReadableSpan): void {
    this.spans.push(span);
  }

  forceFlush(): Promise<void> {
    return Promise.resolve();
  }

  shutdown(): Promise<void> {
    this.spans = [];
    this.processSpanIds = new Map();
    return Promise.resolve();
  }

  /**
   * True when this process started the given span itself: a fetched span
   * with this id is a platform echo of a local span, never remote evidence.
   */
  isProcessSpan(traceId: string, spanId: string): boolean {
    return this.processSpanIds.get(traceId)?.has(spanId) ?? false;
  }

  /**
   * Removes one span by id. Used by the remote trace fetcher to retract a
   * synthetic error span when the judge waits once more and the trace
   * settles after all.
   */
  removeSpanById(spanId: string): void {
    this.spans = this.spans.filter((s) => s.spanContext().spanId !== spanId);
  }

  /**
   * Removes all spans associated with a specific thread.
   * Call this after a scenario run completes to prevent memory growth
   * in long-lived processes.
   * @param threadId - The thread identifier whose spans should be cleared
   */
  clearSpansForThread(threadId: string): void {
    const threadSpans = this.getSpansForThread(threadId);
    const threadSpanIds = new Set(
      threadSpans.map((s) => s.spanContext().spanId)
    );
    for (const span of threadSpans) {
      this.processSpanIds.delete(span.spanContext().traceId);
    }
    this.spans = this.spans.filter(
      (s) => !threadSpanIds.has(s.spanContext().spanId)
    );
  }

  /**
   * Retrieves all spans associated with a specific thread.
   * @param threadId - The thread identifier to filter spans by
   * @returns Array of spans for the given thread
   */
  getSpansForThread(threadId: string): ReadableSpan[] {
    const spanMap = new Map<string, ReadableSpan>();

    // Index all spans by ID
    for (const span of this.spans) {
      spanMap.set(span.spanContext().spanId, span);
    }

    // Check if span or any ancestor belongs to thread
    const belongsToThread = (span: ReadableSpan, visited = new Set<string>()): boolean => {
      const spanId = span.spanContext().spanId;
      if (visited.has(spanId)) return false;
      visited.add(spanId);

      if (span.attributes[attributes.ATTR_LANGWATCH_THREAD_ID] === threadId) {
        return true;
      }
      const parentId = getParentSpanId(span);
      const parentSpan = parentId ? spanMap.get(parentId) : undefined;
      if (parentSpan) {
        return belongsToThread(parentSpan, visited);
      }
      return false;
    };

    return this.spans.filter((span) => belongsToThread(span));
  }
}

/**
 * Extracts the parent span ID from a ReadableSpan. The OTel SDK exposes the
 * parent as a typed SpanContext (parentSpanContext); older span implementations
 * exposed a flat parentSpanId string, which is handled as a fallback.
 */
function getParentSpanId(span: ReadableSpan): string | undefined {
  // The span tree can contain ReadableSpan instances from different OTel SDK
  // versions, so read both shapes via a cast rather than relying on either
  // field being present on the statically-resolved type.
  const s = span as unknown as {
    parentSpanContext?: { spanId?: string };
    parentSpanId?: string;
  };
  return s.parentSpanContext?.spanId ?? s.parentSpanId;
}

/**
 * Singleton instance of the judge span collector.
 */
export const judgeSpanCollector = new JudgeSpanCollector();

File: python/scenario/_tracing/judge_span_collector.py
"""
Collects OpenTelemetry spans for judge evaluation.

Implements SpanProcessor to intercept spans as they complete,
storing them for later retrieval by thread ID.
"""

from typing import Iterable, List, Dict, Optional, Set
from opentelemetry.context import Context
from opentelemetry.sdk.trace import SpanProcessor, ReadableSpan
from langwatch.attributes import AttributeKey


class JudgeSpanCollector(SpanProcessor):
    """
    Collects OpenTelemetry spans for judge evaluation.

    Implements SpanProcessor to intercept spans as they complete.
    Spans can be retrieved by thread ID for inclusion in judge prompts.
    """

    def __init__(self) -> None:
        self._spans: List[ReadableSpan] = []
        # Every span id this process ever STARTED, keyed by trace id. The
        # remote trace fetcher uses it to recognize the scenario process's
        # own spans when the platform echoes them back. The per-thread view
        # below cannot serve that: it walks ancestor attributes, and the walk
        # breaks on spans whose ancestor chain crosses a still-open span (the
        # current turn's spans) or whose instrumentation never tags the
        # thread id (the judge's own model calls via instrumented SDKs).
        self._process_span_ids: Dict[int, Set[int]] = {}
        # Trace ids the remote trace fetcher settled, per thread. Clearing a
        # thread by ended-span discovery cannot reach every registry entry:
        # a fetched trace's local echoes may never end, or never associate
        # with the thread (an instrumented SDK's model call carries no
        # thread id). The fetcher claims the trace ids it touches, and
        # clear_spans_for_thread releases the claims with the thread.
        self._thread_trace_ids: Dict[str, Set[int]] = {}

    def on_start(
        self,
        span: ReadableSpan,
        parent_context: Optional[Context] = None,
    ) -> None:
        """Called when a span starts. Registers it as a process span."""
        span_ctx = span.get_span_context()
        if span_ctx:
            self._process_span_ids.setdefault(span_ctx.trace_id, set()).add(
                span_ctx.span_id
            )

    def on_end(self, span: ReadableSpan) -> None:
        """Called when a span ends. Stores the span for later retrieval."""
        self._spans.append(span)

    def shutdown(self) -> None:
        """Shuts down the processor, clearing all stored spans."""
        self._spans = []
        self._process_span_ids = {}
        self._thread_trace_ids = {}

    def claim_traces(self, thread_id: str, trace_ids: Iterable[int]) -> None:
        """Marks the given trace ids as owned by a thread.

        Lets ``clear_spans_for_thread`` release their process-span registry
        entries even when no ended span ties the trace to the thread.
        """
        self._thread_trace_ids.setdefault(thread_id, set()).update(trace_ids)

    def force_flush(self, timeout_millis: Optional[int] = None) -> bool:
        """Force flush is a no-op for this collector."""
        return True

    def is_process_span(self, trace_id: int, span_id: int) -> bool:
        """True when this process started the given span itself.

        A fetched span with this id is a platform echo of a local span,
        never remote evidence.
        """
        return span_id in self._process_span_ids.get(trace_id, set())

    def remove_span_by_id(self, span_id: int) -> None:
        """Removes one span by id.

        Used by the remote trace fetcher to retract a synthetic error span
        when the judge waits once more and the trace settles after all.
        """
        self._spans = [
            s
            for s in self._spans
            if (ctx.span_id if (ctx := s.get_span_context()) else 0) != span_id
        ]

    def get_spans_for_thread(self, thread_id: str) -> List[ReadableSpan]:
        """
        Retrieves all spans associated with a specific thread.

        Traverses parent relationships to find spans belonging to a thread,
        even if the thread ID is only set on an ancestor span.

        Args:
            thread_id: The thread identifier to filter spans by

        Returns:
            List of spans for the given thread
        """
        span_map: Dict[int, ReadableSpan] = {}

        # Index all spans by ID
        for span in self._spans:
            span_ctx = span.get_span_context()
            span_id = span_ctx.span_id if span_ctx else 0
            span_map[span_id] = span

        return [
            s
            for s in self._spans
            if self._belongs_to_thread(s, thread_id, span_map, set())
        ]

    def _belongs_to_thread(
        self,
        span: ReadableSpan,
        thread_id: str,
        span_map: Dict[int, ReadableSpan],
        visited: set,
    ) -> bool:
        """Check if span or any ancestor belongs to thread.

        Uses a visited set to protect against cycles in parent chains.
        """
        span_ctx = span.get_span_context()
        span_id = span_ctx.span_id if span_ctx else 0
        if span_id in visited:
            return False
        visited.add(span_id)

        attrs = span.attributes or {}
        if attrs.get(AttributeKey.LangWatchThreadId) == thread_id:
            return True

        parent_ctx = span.parent
        if parent_ctx is not None:
            parent_id = parent_ctx.span_id
            if parent_id in span_map:
                return self._belongs_to_thread(
                    span_map[parent_id], thread_id, span_map, visited
                )

        return False

    def clear_spans_for_thread(self, thread_id: str) -> None:
        """Remove all spans associated with a specific thread.

        Called after a scenario run completes to prevent memory buildup
        in long-lived processes.
        """
        span_map: Dict[int, ReadableSpan] = {}
        for span in self._spans:
            span_ctx = span.get_span_context()
            span_id = span_ctx.span_id if span_ctx else 0
            span_map[span_id] = span

        thread_span_ids = set()
        for span in self._spans:
            if self._belongs_to_thread(span, thread_id, span_map, set()):
                span_ctx = span.get_span_context()
                if span_ctx:
                    thread_span_ids.add(span_ctx.span_id)
                    self._process_span_ids.pop(span_ctx.trace_id, None)

        for trace_id in self._thread_trace_ids.pop(thread_id, set()):
            self._process_span_ids.pop(trace_id, None)

        self._spans = [
            s
            for s in self._spans
            if (ctx.span_id if (ctx := s.get_span_context()) else 0)
            not in thread_span_ids
        ]


# Singleton instance
judge_span_collector = JudgeSpanCollector()
